catalog/dags/maintenance/pr_review_reminders/pr_review_reminders.py

```diff
@@ -173,38 +173,42 @@ def base_repo_name(pr: dict):
     return pr["base"]["repo"]["name"]
 
 
-_BRANCH_PROTECTION_CACHE = defaultdict(dict)
+def get_branch_protection(
+    gh: GitHubAPI, repo: str, branch_name: str, cache: dict
+) -> dict:
+    if branch_name not in cache[repo]:
+        cache[repo][branch_name] = gh.get_branch_protection(repo, branch_name)
 
-
-def get_branch_protection(gh: GitHubAPI, repo: str, branch_name: str) -> dict:
-    if branch_name not in _BRANCH_PROTECTION_CACHE[repo]:
-        _BRANCH_PROTECTION_CACHE[repo][branch_name] = gh.get_branch_protection(
-            repo, branch_name
-        )
-
-    return _BRANCH_PROTECTION_CACHE[repo][branch_name]
+    return cache[repo][branch_name]
 
 
 def get_min_required_approvals(gh: GitHubAPI, pr: dict) -> int:
+    branch_protection_cache = defaultdict(dict)
```

Comment:
This doesn't resolve the issue, unfortunately, at least not the detail I mentioned about using a cache per-DAG run for purposes of consistency within the run. Maybe that isn't important, but this also doesn't persist the cache within the DAG run.

`get_min_required_approvals` is called in a loop over all `urgent_prs` in the main task function. This cache is recreated for each of those PRs within the context of this function, and basically doesn't do any caching. Each time this function runs, the cache will be empty, and it will fetch the branch protections for the queries tested in each of the calls to `get_branch_protection` within this function each time, each of which use different branches for the query, and so do not re-use cached values.

Solving this problem requires a wider refactor of this DAG so that the cache can be shared within the DAG run but ephemeral outside of it. The easiest way that comes to mind to do that is to move these individual functions and the body of the task function into methods on a class, and store the cache as an instance attribute of the class. Then the task function should create a new instance of the class, and call whatever method encapsulates the logic currently in the task function.

That's short of re-writing this into multiple tasks with distributed state (including the cache).

Reply:
Gah, okay thanks for expounding on this. I didn't quite think through the steps when making the change here. Let me take another look at it and try to retain the cache across the DAG.

Reply:
There's actually a simpler solution here which is just to move the cache one more level up! A small change within the primary function which doesn't require any significant refactoring or changes to tests. I added some logging locally to ensure the cache is being used appropriately, and it was 😄 Going with that approach for now!

Reply:
Cool! I hadn't looked at it a lot and was worried it would require passing through to lots of functions (like we do with the `gh` instance), but glad it was that easy :grin: